swarm/storage/pyramid.go

```diff
@@ -73,128 +130,496 @@ func NewPyramidChunker(params *ChunkerParams) (self *PyramidChunker) {
+
+	chunkWG.Wait()
+	self.enqueueTreeChunk(chunkLevel, ent, chunkWG, jobC, quitC, last)
+
+	compress := false
+	endLvl := self.branches
+	for lvl := int64(0); lvl < self.branches; lvl++ {
+		lvlCount := int64(len(chunkLevel[lvl]))
+		if lvlCount >= self.branches {
+			endLvl = lvl + 1
+			compress = true
+			break
 		}
-		pend.Done()
 	}
+
+	if compress == false && last == false {
+		return
+	}
+
+	// Wait for all the keys to be processed before compressing the tree
+	chunkWG.Wait()
+
+	for lvl := int64(ent.level); lvl < endLvl; lvl++ {
+
+		lvlCount := int64(len(chunkLevel[lvl]))
+		if lvlCount == 1 && last == true {
+			copy(rootKey, chunkLevel[lvl][0].key)
+			return
+		}
+
+		for startCount := int64(0); startCount < lvlCount; startCount += self.branches {
+
+			endCount := startCount + self.branches
+			if endCount > lvlCount {
+				endCount = lvlCount
+			}
+
+			var nextLvlCount int64
+			var tempEntry *TreeEntry
+			if len(chunkLevel[lvl+1]) > 0 {
+				nextLvlCount = int64(len(chunkLevel[lvl+1]) - 1)
+				tempEntry = chunkLevel[lvl+1][nextLvlCount]
+			}
+			if isAppend == true && tempEntry != nil && tempEntry.updatePending == true {
+				updateEntry := &TreeEntry{
+					level:         int(lvl + 1),
+					branchCount:   0,
+					subtreeSize:   0,
+					chunk:         make([]byte, self.chunkSize+8),
+					key:           make([]byte, self.hashSize),
+					index:         int(nextLvlCount),
+					updatePending: true,
+				}
+				for index := int64(0); index < lvlCount; index++ {
+					updateEntry.branchCount++
+					updateEntry.subtreeSize += chunkLevel[lvl][index].subtreeSize
+					copy(updateEntry.chunk[8+(index*self.hashSize):8+((index+1)*self.hashSize)], chunkLevel[lvl][index].key[:self.hashSize])
+				}
+
+				self.enqueueTreeChunk(chunkLevel, updateEntry, chunkWG, jobC, quitC, last)
+
+			} else {
+
+				noOfBranches := endCount - startCount
+				newEntry := &TreeEntry{
+					level:         int(lvl + 1),
+					branchCount:   noOfBranches,
+					subtreeSize:   0,
+					chunk:         make([]byte, (noOfBranches*self.hashSize)+8),
+					key:           make([]byte, self.hashSize),
+					index:         int(nextLvlCount),
+					updatePending: false,
+				}
+
+				index := int64(0)
+				for i := startCount; i < endCount; i++ {
+					entry := chunkLevel[lvl][i]
+					newEntry.subtreeSize += entry.subtreeSize
+					copy(newEntry.chunk[8+(index*self.hashSize):8+((index+1)*self.hashSize)], entry.key[:self.hashSize])
+					index++
+				}
+
+				self.enqueueTreeChunk(chunkLevel, newEntry, chunkWG, jobC, quitC, last)
+
+			}
+
+		}
+
+		if isAppend == false {
+			chunkWG.Wait()
+			if compress == true {
+				chunkLevel[lvl] = nil
+			}
+		}
+	}
+
+}
+
+func (self *PyramidChunker) enqueueTreeChunk(chunkLevel [][]*TreeEntry, ent *TreeEntry, chunkWG *sync.WaitGroup, jobC chan *chunkJob, quitC chan bool, last bool) {
+
+	if ent != nil {
```

Comment:
How about just not calling the function if this is `nil`?

Reply:
enqueueTreeChunk is called in multiple places.. instead of putting the condition in multiple places.. i though it would be nice to have it in one place. :-)